common/model/transfer.c

```diff
@@ -1,676 +1,474 @@
+  size_t flex_ret = 0;
+
+  if (transfer->type == VALUE_OUT) {
+    output_info = (transfer_value_out_t*)transfer->meta;
+    tx = transfer_iterator->transaction;
+    if (transfer_iterator->current_transfer_transaction_index == 0) {
+      if (transfer_iterator->transaction_signature) {
+        free(transfer_iterator->transaction_signature);
+      }
+      iota_signature_generator iota_signature_gen =
+          transfer_iterator->iota_signature_gen;
+      transfer_iterator->transaction_signature =
+          (flex_trit_t*)iota_signature_gen(
+              output_info->seed, output_info->seed_index, output_info->security,
+              transfer_iterator->bundle_hash);
+      tx->value = transfer->value;
+    }
+    flex_ret = flex_trits_slice(
+        tx->signature_or_message, NUM_TRITS_SIGNATURE,
+        transfer_iterator->transaction_signature,
+        NUM_TRITS_SIGNATURE * output_info->security,
+        NUM_TRITS_SIGNATURE *
+            transfer_iterator->current_transfer_transaction_index,
+        NUM_TRITS_SIGNATURE);
+    if (flex_ret == 0) {
+      log_warning(TRANSFER_LOGGER_ID, "[%s:%d] flex_trits slicing failed.\n",
+                  __func__, __LINE__);
+    }
+  } else {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] transfer type not match.\n",
+              __func__, __LINE__);
+  }
+}
+
+static void transfer_iterator_next_input_transaction(
+    transfer_iterator_t* transfer_iterator, transfer_t* transfer) {
+  transfer_value_in_t* value_in = NULL;
+  iota_transaction_t tx = transfer_iterator->transaction;
+
+  if (transfer->type == VALUE_IN) {
+    value_in = (transfer_value_in_t*)transfer->meta;
+    memcpy(tx->signature_or_message, value_in->data, sizeof(value_in->len));
+    tx->value = transfer->value;
+  } else {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] transfer type not match.\n",
+              __func__, __LINE__);
+  }
+}
+
+/***********************************************************************************************************
+ * Public interface
+ ***********************************************************************************************************/
+
+void transfer_logger_init() {
+  logger_helper_init(TRANSFER_LOGGER_ID, LOGGER_DEBUG, true);
+  log_info(TRANSFER_LOGGER_ID, "Enable logger %s.\n", TRANSFER_LOGGER_ID);
+}
+
+void transfer_logger_destroy() {
+  log_info(TRANSFER_LOGGER_ID, "Destroy logger %s.\n", TRANSFER_LOGGER_ID);
+  logger_helper_destroy(TRANSFER_LOGGER_ID);
+}
+
+bool validate_output(transfer_value_out_t const* const output) {
+  if (!output->seed) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] seed is NULL.\n", __func__,
+              __LINE__);
+    return false;
+  }
+
+  if (output->security < 1 || output->security > 3) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] security leve error.\n", __func__,
+              __LINE__);
+    return false;
+  }
+  return true;
+}
+
+// create data transfer object.
+transfer_t* transfer_data_new(flex_trit_t const* const address,
```

Comment:
What about `_init` functions that takes and fills a struct instead of `_new` functions that allocs and fills a struct ?

Reply:
it allocates transfer_t.

Reply:
Yes, but why ? Is this really necessary ?
By taking a `transfer_t *` as parameter we allow the user to fill a static or a dynamic `transfer_t`. IMHO we should avoid dynamic allocation as much as possible in code used in the cclient.

Reply:
this change just provides helper methods and keeps original design in this model. 
In my understanding, structures are mostly pointers in the transfer.

Reply:
Yes but since it will only be used by the client maybe it's a good thing to adapt it to the client needs. Pointers can be addresses of static structures. We'll probably need to remove dynamic allocation someday.